scylla/src/transport/load_balancing/latency_aware.rs

```diff
@@ -0,0 +1,295 @@
+
+        let ask_for_nodes_clone = ask_for_nodes.clone();
+        let min_latency_clone = min_latency.clone();
+
+        let (updater_fut, updater_handle) = async move {
+            loop {
+                ask_for_nodes.store(true, Ordering::Relaxed);
+                update_scheduler.tick().await;
+                let nodes = nodes_rx.latest();
+                if let Some(nodes) = nodes {
+                    if nodes.is_empty() {
+                        error!("Empty node list - driver/policy bug!");
+                        return;
+                    }
+                    trace!("LatencyAwarePolicy: updating min average");
+
+                    let min_avg = nodes
+                        .iter()
+                        .filter_map(|node| {
+                            node.average_latency
+                                .read()
+                                .unwrap()
+                                .map(|timestamped_average| timestamped_average.average)
+                        })
+                        .min();
+                    if let Some(min_avg) = min_avg {
+                        min_latency.store(min_avg);
+                    }
+                }
+            }
+        }
+        .remote_handle();
+        tokio::task::spawn(updater_fut);
+
+        Self {
+            exclusion_threshold,
+            retry_period,
+            update_rate,
+            minimum_measurements,
+            last_min_latency: min_latency_clone,
+            nodes_tx,
+            ask_for_nodes: ask_for_nodes_clone,
+            _updater_handle: updater_handle,
+        }
+    }
+
+    fn refresh_last_min_avg_nodes(&self, nodes: &[Arc<Node>]) {
+        if self.ask_for_nodes.swap(false, Ordering::Relaxed) {
+            let _ = self.nodes_tx.update(Some(nodes.to_owned()));
+            trace!("LatencyAwarePolicy: updating nodes list for the min average updater");
+        }
+    }
+
+    fn make_plan<'a>(&self, plan: Vec<Arc<Node>>) -> Plan<'a> {
+        self.refresh_last_min_avg_nodes(&plan);
+        let min_avg_latency = match self.last_min_latency.load() {
+            Some(min_avg) => min_avg,
+            None => return Box::new(plan.into_iter()), // noop, as no latency data has been collected yet
+        };
+
+        Box::new(IteratorWithSkippedNodes::new(
+            Box::new(plan.into_iter()),
+            self.exclusion_threshold,
+            self.retry_period,
+            self.minimum_measurements,
+            min_avg_latency,
+        ))
+    }
+}
+
+impl Default for LatencyAwarePolicy {
+    fn default() -> Self {
+        Self::new(
+            2_f64,
+            Duration::from_secs(10),
+            Duration::from_millis(100),
+            50,
+        )
+    }
+}
+
+impl LoadBalancingPolicy for LatencyAwarePolicy {
+    fn plan<'a>(&self, _statement: &Statement, cluster: &'a ClusterData) -> Plan<'a> {
+        self.make_plan(cluster.all_nodes.clone())
```

Comment:
IIUC, Java driver's implementation behaves the same: it does not shuffle the replicas on its own. The idea is that one should use a round robin flavour of this policy for shifting the "fast" and the "slow" replica sets with each query. Perhaps it is a good idea to add such a policy, RoundRobinLatencyAware policy? And maybe even RoundRobinLatencyAndTokenAware policy?

Reply:
I'd rather not add every non-empty combination of {TokenAware, LatencyAware, DCAware, RoundRobin} to the driver. We can take one of the two approaches to combat the exponential grow:

- Implement the LatencyAwarePolicy as a "wrapper" policy. It will contain a child policy and delegate some details to it using the existing ChildLoadBalancingPolicy trait. We already to it for TokenAware policy and DCAwarePolicy, I think. On the other hand, it's a bit hard to think about how the policy will actually behave when composed. We are moving away from this approach in the load balancing policy refactor which was worked on by @havaker / @cvybhu (https://github.com/scylladb/scylla-rust-driver/pull/491).
- Implement a LatencyAwarePolicy which supports all of the features (token-awareness, DC awareness, round robin) but allows to switch them off in runtime with boolean flags.

I'm not sure which approach is better. @cvybhu thoughts?

Reply:
I've implemented the first approach. Looking forward for reviews. @piodul @cvybhu